{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process all the data using TPP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "******** Run 1: input/HJKK5BCX2_TCGGAA_1.fastq ********\n",
      "******** TPP finished in 0 seconds! Average iteration time over 1 iterations:  0 seconds. ********\n",
      "# title: Tn-Seq Pre-Processor\n",
      "# date: 08/16/2020 16:46:03\n",
      "# command: python /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/tpp -himar1 -bwa /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa -bwa-alg aln -ref input/Genome/MAC109.fa -replicon-ids CP029332,CP029333,CP029334 -reads1 input/HJKK5BCX2_TCGGAA_1.fastq -reads2 input/HJKK5BCX2_TCGGAA_2.fastq -window-size 6 -primer AACCTGTTA -mismatches 2 -output output/HJKK5BCX2_TCGGAA\n",
      "# transposon type: Himar1\n",
      "# protocol type: Sassetti\n",
      "# bwa flags:\n",
      "# read1: input/HJKK5BCX2_TCGGAA_1.fastq\n",
      "# read2: input/HJKK5BCX2_TCGGAA_2.fastq\n",
      "# ref_genome: input/Genome/MAC109.fa\n",
      "# replicon_ids: CP029332,CP029333,CP029334\n",
      "# total_reads (or read pairs): 5185840\n",
      "# trimmed_reads (reads with valid Tn prefix, and insert size>20bp): 4739037\n",
      "# reads1_mapped: 4471310\n",
      "# reads2_mapped: 4359462\n",
      "# mapped_reads (both R1 and R2 map into genome, and R2 has a proper barcode): 4099024\n",
      "# read_count (TA sites only, for Himar1):\n",
      "#   CP029332: 3650591\n",
      "#   CP029333: 334722\n",
      "#   CP029334: 86719\n",
      "# template_count:\n",
      "#   CP029332: 2946500\n",
      "#   CP029333: 270433\n",
      "#   CP029334: 71297\n",
      "# template_ratio (reads per template):\n",
      "#   CP029332: 1.24\n",
      "#   CP029333: 1.24\n",
      "#   CP029334: 1.22\n",
      "# TA_sites:\n",
      "#   CP029332: 57588\n",
      "#   CP029333: 2255\n",
      "#   CP029334: 286\n",
      "# TAs_hit:\n",
      "#   CP029332: 40401\n",
      "#   CP029333: 1932\n",
      "#   CP029334: 242\n",
      "# density:\n",
      "#   CP029332: 0.702\n",
      "#   CP029333: 0.857\n",
      "#   CP029334: 0.846\n",
      "# max_count (among templates):\n",
      "#   CP029332: 2293\n",
      "#   CP029333: 1613\n",
      "#   CP029334: 30889\n",
      "# max_site (coordinate):\n",
      "#   CP029332: 2118281\n",
      "#   CP029333: 54501\n",
      "#   CP029334: 9246\n",
      "# NZ_mean (among templates):\n",
      "#   CP029332: 72.9\n",
      "#   CP029333: 140.0\n",
      "#   CP029334: 294.6\n",
      "# FR_corr (Fwd templates vs. Rev templates):\n",
      "#   CP029332: 0.866\n",
      "#   CP029333: 0.859\n",
      "#   CP029334: 0.998\n",
      "# BC_corr (reads vs. templates, summed over both strands):\n",
      "#   CP029332: 0.999\n",
      "#   CP029333: 0.999\n",
      "#   CP029334: 1.000\n",
      "# primer_matches: 17 reads (0.0%) contain CTAGAGGGCCCAATTCGCCCTATAGTGAGT (Himar1)\n",
      "# vector_matches: 346 reads (0.0%) contain CTAGACCGTCCAGTCTGGCAGGCCGGAAAC (phiMycoMarT7)\n",
      "# adapter_matches: 537 reads (0.0%) contain GATCGGAAGAGCACACGTCTGAACTCCAGTCAC (Illumina/TruSeq index)\n",
      "# misprimed_reads: 215768 reads (4.2%) contain Himar1 prefix but don't end in TGTTA\n",
      "# read_length: 75 bp\n",
      "# mean_R1_genomic_length: 47.1 bp\n",
      "# mean_R2_genomic_length: 26.4 bp\n",
      "Creating prefixes file with all prefixes from all runs...\n",
      "Created 'output/input_prefixes.txt'.\n",
      "\n",
      "Creating CSV file with all samples processed by TPP...\n",
      "[+] scripts/wig_gb_to_csv.py starting with parameters:\n",
      "....Wiggle prefix(es)         output/HJKK5BCX2_TCGGAA\n",
      "....Record ID(s)              CP029332, CP029333, CP029334\n",
      "....GenBank file:             input/Genome/MAC109.gb\n",
      "....Unique identifier fields: locus_tag, record_id_start_end_strand\n",
      "....Fields:                   product, regulatory_class, bound_moiety\n",
      "....Output file:              output/input.csv\n",
      "[+] Combining Wiggles with prefix '['output/HJKK5BCX2_TCGGAA']' with Record ID 'CP029332' from 'input/Genome/MAC109.gb'...\n",
      "[+] Finished processing Record ID 'CP029332' successfully in 5.994068 seconds.\n",
      "[+] Combining Wiggles with prefix '['output/HJKK5BCX2_TCGGAA']' with Record ID 'CP029333' from 'input/Genome/MAC109.gb'...\n",
      "[+] Finished processing Record ID 'CP029333' successfully in 1.035354 seconds.\n",
      "[+] Combining Wiggles with prefix '['output/HJKK5BCX2_TCGGAA']' with Record ID 'CP029334' from 'input/Genome/MAC109.gb'...\n",
      "[+] Finished processing Record ID 'CP029334' successfully in 0.887105 seconds.\n",
      "[+] Finished processing all records successfully in 7.916609 seconds.\n",
      "[+] Writing CSV files...\n",
      "[+] 'output/input.csv' written.\n",
      "[+] Success. Exiting...\n",
      "Created 'output/input.csv'.\n",
      "\n",
      "********** TPP driver script finished in a total of 0 seconds **********\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[tn_preprocess] running pre-processing on input/HJKK5BCX2_TCGGAA_1.fastq\n",
      "[tn_preprocess] protocol: Sassetti\n",
      "[tn_preprocess] transposon type: Himar1\n",
      "[tn_preprocess] One reference genome specified: input/Genome/MAC109.fa, containing 3 records.\n",
      "[tn_preprocess] extracting reads...\n",
      "[tn_preprocess] fastq2reads: input/HJKK5BCX2_TCGGAA_1.fastq -> output/HJKK5BCX2_TCGGAA.reads1\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] fastq2reads: input/HJKK5BCX2_TCGGAA_2.fastq -> output/HJKK5BCX2_TCGGAA.reads2\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] fixing headers of paired reads for bwa...\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] extracting barcodes and genomic parts of reads...\n",
      "[tn_preprocess] creating output/HJKK5BCX2_TCGGAA.trimmed1\n",
      "[tn_preprocess] prefix sequence: AACCTGTTA\n",
      "[tn_preprocess] Looking for start of Tn prefix with P,Q = 16,22 (origin = 19, window size = 6)\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] 5000000 reads processed\n",
      "[tn_preprocess] creating output/HJKK5BCX2_TCGGAA.trimmed2\n",
      "[tn_preprocess] creating output/HJKK5BCX2_TCGGAA.barcodes2\n",
      "[tn_preprocess] creating output/HJKK5BCX2_TCGGAA.genomic2\n",
      "[tn_preprocess] 1000000 reads processed\n",
      "[tn_preprocess] 2000000 reads processed\n",
      "[tn_preprocess] 3000000 reads processed\n",
      "[tn_preprocess] 4000000 reads processed\n",
      "[tn_preprocess] creating output/HJKK5BCX2_TCGGAA.barcodes1\n",
      "[tn_preprocess] mapping reads using BWA...(this takes a couple of minutes)\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.trimmed1 > output/HJKK5BCX2_TCGGAA.sai1\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 8.84 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 9.25 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.84 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.71 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 9.52 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.36 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.44 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.61 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.95 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.06 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.48 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2883584 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.74 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3145728 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.31 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3407872 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.85 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 3670016 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.18 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3932160 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.46 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 4194304 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 7.55 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 4456448 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 8.62 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 4718592 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 0.51 sec\n",
      "[bwa_aln_core] write to the disk... 0.00 sec\n",
      "[bwa_aln_core] 4739037 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.trimmed1\n",
      "[main] Real time: 151.227 sec; CPU: 150.155 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.genomic2 > output/HJKK5BCX2_TCGGAA.sai2\n",
      "[bwa_aln] 17bp reads: max_diff = 2\n",
      "[bwa_aln] 38bp reads: max_diff = 3\n",
      "[bwa_aln] 64bp reads: max_diff = 4\n",
      "[bwa_aln] 93bp reads: max_diff = 5\n",
      "[bwa_aln] 124bp reads: max_diff = 6\n",
      "[bwa_aln] 157bp reads: max_diff = 7\n",
      "[bwa_aln] 190bp reads: max_diff = 8\n",
      "[bwa_aln] 225bp reads: max_diff = 9\n",
      "[bwa_aln_core] calculate SA coordinate... 5.13 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 262144 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.81 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 524288 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.84 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 786432 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.70 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1048576 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.48 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1310720 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.48 sec\n",
      "[bwa_aln_core] write to the disk... 0.02 sec\n",
      "[bwa_aln_core] 1572864 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.90 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 1835008 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.88 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2097152 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.64 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2359296 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.92 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 2621440 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.70 sec\n",
      "[bwa_aln_core] write to the disk... 0.04 sec\n",
      "[bwa_aln_core] 2883584 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.08 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3145728 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.29 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3407872 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.90 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3670016 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 5.02 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 3932160 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.95 sec\n",
      "[bwa_aln_core] write to the disk... 0.04 sec\n",
      "[bwa_aln_core] 4194304 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.96 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 4456448 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 4.69 sec\n",
      "[bwa_aln_core] write to the disk... 0.03 sec\n",
      "[bwa_aln_core] 4718592 sequences have been processed.\n",
      "[bwa_aln_core] calculate SA coordinate... 0.38 sec\n",
      "[bwa_aln_core] write to the disk... 0.00 sec\n",
      "[bwa_aln_core] 4739037 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa aln input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.genomic2\n",
      "[main] Real time: 93.299 sec; CPU: 92.205 sec\n",
      "\n",
      "[tn_preprocess] /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.sai1 output/HJKK5BCX2_TCGGAA.sai2 output/HJKK5BCX2_TCGGAA.trimmed1 output/HJKK5BCX2_TCGGAA.genomic2 > output/HJKK5BCX2_TCGGAA.sam\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (144, 212, 274)\n",
      "[infer_isize] low and high boundaries: 50 and 534 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 210069 pairs: 217.571 +/- 89.810\n",
      "[infer_isize] skewness: 0.340; kurtosis: -0.113; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 752 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 2.41 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3854 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3330 out of 16479 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 9 out of 9487 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.91 sec\n",
      "[bwa_sai2sam_pe_core] 262144 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (138, 207, 269)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 207272 pairs: 213.238 +/- 89.222\n",
      "[infer_isize] skewness: 0.345; kurtosis: -0.151; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 744 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 2.89 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3850 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3485 out of 17863 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 12 out of 9759 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.17 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.06 sec\n",
      "[bwa_sai2sam_pe_core] 524288 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (140, 209, 271)\n",
      "[infer_isize] low and high boundaries: 50 and 533 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 203073 pairs: 215.248 +/- 89.497\n",
      "[infer_isize] skewness: 0.340; kurtosis: -0.135; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 748 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.45 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3721 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3481 out of 22298 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 7 out of 9136 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.38 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.00 sec\n",
      "[bwa_sai2sam_pe_core] 786432 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (142, 211, 273)\n",
      "[infer_isize] low and high boundaries: 50 and 535 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 209209 pairs: 216.415 +/- 89.746\n",
      "[infer_isize] skewness: 0.341; kurtosis: -0.124; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 750 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.30 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3923 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3288 out of 16852 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 12 out of 9467 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.35 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.01 sec\n",
      "[bwa_sai2sam_pe_core] 1048576 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (135, 205, 267)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 207377 pairs: 211.775 +/- 89.184\n",
      "[infer_isize] skewness: 0.358; kurtosis: -0.131; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 742 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.20 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3783 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3637 out of 17366 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 13 out of 10372 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.38 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.19 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.97 sec\n",
      "[bwa_sai2sam_pe_core] 1310720 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (143, 212, 273)\n",
      "[infer_isize] low and high boundaries: 50 and 533 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 209818 pairs: 217.317 +/- 89.594\n",
      "[infer_isize] skewness: 0.333; kurtosis: -0.121; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 750 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 2.78 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3887 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3658 out of 16383 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 14 out of 9335 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.35 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.94 sec\n",
      "[bwa_sai2sam_pe_core] 1572864 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (141, 210, 271)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 206180 pairs: 215.267 +/- 89.359\n",
      "[infer_isize] skewness: 0.337; kurtosis: -0.132; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 747 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.66 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3805 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 4853 out of 18436 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 9 out of 9185 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.37 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.01 sec\n",
      "[bwa_sai2sam_pe_core] 1835008 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (144, 212, 274)\n",
      "[infer_isize] low and high boundaries: 50 and 534 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 209941 pairs: 217.690 +/- 89.629\n",
      "[infer_isize] skewness: 0.336; kurtosis: -0.116; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 751 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.33 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3930 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3650 out of 16427 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 7 out of 9225 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.35 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.04 sec\n",
      "[bwa_sai2sam_pe_core] 2097152 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (139, 208, 269)\n",
      "[infer_isize] low and high boundaries: 50 and 529 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 207554 pairs: 213.733 +/- 89.273\n",
      "[infer_isize] skewness: 0.349; kurtosis: -0.121; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 745 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.15 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3795 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3538 out of 17169 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 11 out of 9838 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.41 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.97 sec\n",
      "[bwa_sai2sam_pe_core] 2359296 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (143, 212, 273)\n",
      "[infer_isize] low and high boundaries: 50 and 533 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 210527 pairs: 217.273 +/- 89.425\n",
      "[infer_isize] skewness: 0.322; kurtosis: -0.140; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 749 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.69 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3918 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3420 out of 16450 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 17 out of 9325 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.35 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.93 sec\n",
      "[bwa_sai2sam_pe_core] 2621440 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (139, 208, 270)\n",
      "[infer_isize] low and high boundaries: 50 and 532 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 208398 pairs: 214.270 +/- 89.299\n",
      "[infer_isize] skewness: 0.346; kurtosis: -0.132; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 746 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 2.47 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3826 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3249 out of 17011 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 6 out of 9844 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.02 sec\n",
      "[bwa_sai2sam_pe_core] 2883584 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (141, 210, 271)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 208936 pairs: 215.470 +/- 89.491\n",
      "[infer_isize] skewness: 0.339; kurtosis: -0.141; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 748 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.06 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3693 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3331 out of 16981 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 12 out of 9500 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.34 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.97 sec\n",
      "[bwa_sai2sam_pe_core] 3145728 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (140, 210, 272)\n",
      "[infer_isize] low and high boundaries: 50 and 536 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 210779 pairs: 215.711 +/- 89.709\n",
      "[infer_isize] skewness: 0.338; kurtosis: -0.129; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 749 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.04 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3846 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 2931 out of 15658 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 12 out of 9679 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.34 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 1.06 sec\n",
      "[bwa_sai2sam_pe_core] 3407872 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (138, 207, 269)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 206638 pairs: 213.648 +/- 89.366\n",
      "[infer_isize] skewness: 0.353; kurtosis: -0.127; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 745 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 2.99 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3743 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3420 out of 17627 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 9 out of 10000 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.36 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.15 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.94 sec\n",
      "[bwa_sai2sam_pe_core] 3670016 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (143, 212, 273)\n",
      "[infer_isize] low and high boundaries: 50 and 533 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 209669 pairs: 217.258 +/- 89.516\n",
      "[infer_isize] skewness: 0.333; kurtosis: -0.123; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 750 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.05 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3827 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3658 out of 16660 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 11 out of 9151 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.34 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.99 sec\n",
      "[bwa_sai2sam_pe_core] 3932160 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (141, 210, 271)\n",
      "[infer_isize] low and high boundaries: 50 and 531 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 205813 pairs: 215.243 +/- 89.083\n",
      "[infer_isize] skewness: 0.331; kurtosis: -0.146; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 745 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.60 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3779 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 4889 out of 18675 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 13 out of 9243 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.37 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.96 sec\n",
      "[bwa_sai2sam_pe_core] 4194304 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (144, 212, 274)\n",
      "[infer_isize] low and high boundaries: 50 and 534 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 211341 pairs: 217.645 +/- 89.786\n",
      "[infer_isize] skewness: 0.335; kurtosis: -0.124; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 752 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.38 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3929 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 3183 out of 15533 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 7 out of 9418 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.37 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.17 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.95 sec\n",
      "[bwa_sai2sam_pe_core] 4456448 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (140, 209, 270)\n",
      "[infer_isize] low and high boundaries: 50 and 530 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 206764 pairs: 214.507 +/- 89.032\n",
      "[infer_isize] skewness: 0.344; kurtosis: -0.123; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 744 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 3.08 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 3869 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 4378 out of 18049 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 9 out of 9459 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.37 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.16 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.95 sec\n",
      "[bwa_sai2sam_pe_core] 4718592 sequences have been processed.\n",
      "[bwa_sai2sam_pe_core] convert to sequence coordinate...\n",
      "[infer_isize] (25, 50, 75) percentile: (141, 209, 272)\n",
      "[infer_isize] low and high boundaries: 50 and 534 for estimating avg and std\n",
      "[infer_isize] inferred external isize from 15984 pairs: 215.105 +/- 89.462\n",
      "[infer_isize] skewness: 0.350; kurtosis: -0.077; ap_prior: 1.00e-05\n",
      "[infer_isize] inferred maximum insert size: 747 (5.95 sigma)\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.27 sec\n",
      "[bwa_sai2sam_pe_core] changing coordinates of 274 alignments.\n",
      "[bwa_sai2sam_pe_core] align unmapped mate...\n",
      "[bwa_paired_sw] 326 out of 1498 Q17 singletons are mated.\n",
      "[bwa_paired_sw] 1 out of 682 Q17 discordant pairs are fixed.\n",
      "[bwa_sai2sam_pe_core] time elapses: 0.03 sec\n",
      "[bwa_sai2sam_pe_core] refine gapped alignments... 0.01 sec\n",
      "[bwa_sai2sam_pe_core] print alignments... 0.08 sec\n",
      "[bwa_sai2sam_pe_core] 4739037 sequences have been processed.\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /home/will/src_and_bin/Other_Software/miniconda3/envs/tnseq_avium_abx_fordistro/bin/bwa sampe input/Genome/MAC109.fa output/HJKK5BCX2_TCGGAA.sai1 output/HJKK5BCX2_TCGGAA.sai2 output/HJKK5BCX2_TCGGAA.trimmed1 output/HJKK5BCX2_TCGGAA.genomic2\n",
      "[main] Real time: 109.284 sec; CPU: 89.269 sec\n",
      "\n",
      "[tn_preprocess] tabulating template counts and statistics for reference genome input/Genome/MAC109.fa\n",
      "[tn_preprocess] writing output/HJKK5BCX2_TCGGAA_CP029332.wig\n",
      "[tn_preprocess] writing output/HJKK5BCX2_TCGGAA_CP029333.wig\n",
      "[tn_preprocess] writing output/HJKK5BCX2_TCGGAA_CP029334.wig\n",
      "[tn_preprocess] writing output/HJKK5BCX2_TCGGAA.tn_stats\n",
      "[tn_preprocess] Done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#SET THESE VARIABLES FOR YOUR LOCAL FILE STRUCTURE:\n",
    "PATH_TO_DATA=input #Note: Any fastq files located here will be processed for alignment with tpp\n",
    "OUT_DIR=output\n",
    "REFS=input/Genome/MAC109.fa\n",
    "GENBANK_FILE=input/Genome/MAC109.gb\n",
    "\n",
    "\n",
    "#Definition of variables for processing\n",
    "PYTHON2=$(which python2)\n",
    "BWA=$(which bwa)\n",
    "BWA_ALG=\"aln\"\n",
    "\n",
    "REPLICON_ID=\"CP029332,CP029333,CP029334\"\n",
    "FASTQ_DIR=$PATH_TO_DATA\n",
    "\n",
    "PREFIXES_OUTFILE=$OUT_DIR/`basename $FASTQ_DIR`_prefixes.txt\n",
    "\n",
    "# These are used for creating a CSV file\n",
    "CSV_OUTFILE=$OUT_DIR/`basename $FASTQ_DIR`.csv\n",
    "UNIQUE_FIELDS=\"locus_tag\"\n",
    "FIELDS=\"product regulatory_class bound_moiety\"\n",
    "\n",
    "#Parameter settings for tpp\n",
    "PRIMER=AACCTGTTA\n",
    "MISMATCHES=2\n",
    "WINDOW_SIZE=6\n",
    "\n",
    "###################################################################\n",
    "\n",
    "#Process raw fastq files using tpp\n",
    "COUNTER=0\n",
    "INITIAL_START_TIME=$SECONDS\n",
    "#for FASTQ in $FASTQ_DIR/*_1.fastq; do\n",
    "for FASTQ in $FASTQ_DIR/*TCGGAA_1.fastq; do\n",
    " (( COUNTER += 1 ))\n",
    " echo \"******** Run $COUNTER: $FASTQ ********\"\n",
    " READS1=$FASTQ\n",
    " READS2=${FASTQ/_1.fastq/_2.fastq}\n",
    "\n",
    " OUTNAME=$(basename $FASTQ)\n",
    " OUTNAME=${OUTNAME/_1.fastq/}\n",
    " ITERATION_START_TIME=$SECONDS\n",
    "  tpp -himar1 -bwa $BWA -bwa-alg $BWA_ALG -ref $REFS -replicon-ids $REPLICON_ID -reads1 $READS1 -reads2 $READS2 -window-size $WINDOW_SIZE -primer $PRIMER -mismatches $MISMATCHES -output $OUT_DIR/$OUTNAME &\n",
    " ITERATION_END_TIME=$SECONDS\n",
    " (( ITERATION_TIME = ITERATION_END_TIME - ITERATION_START_TIME ))\n",
    "\n",
    " (( TOTAL_RUN_TIME = SECONDS - INITIAL_START_TIME )) \n",
    " (( CURRENT_AVG = TOTAL_RUN_TIME / COUNTER ))\n",
    " echo \"******** TPP finished in $ITERATION_TIME seconds! Average iteration time over $COUNTER iterations:  $CURRENT_AVG seconds. ********\"\n",
    "done\n",
    "\n",
    "wait\n",
    "echo \"Creating prefixes file with all prefixes from all runs...\"\n",
    "basename -a $OUT_DIR/*.wig | cut -c-16 | uniq > $PREFIXES_OUTFILE\n",
    "echo \"Created '$PREFIXES_OUTFILE'.\"\n",
    "echo \"\"\n",
    "echo \"Creating CSV file with all samples processed by TPP...\"\n",
    "$PYTHON2 scripts/wig_gb_to_csv.py -l $PREFIXES_OUTFILE -g $GENBANK_FILE -u $UNIQUE_FIELDS -f $FIELDS -o $CSV_OUTFILE\n",
    "echo \"Created '$CSV_OUTFILE'.\"\n",
    "echo \"\"\n",
    "echo \"********** TPP driver script finished in a total of $TOTAL_RUN_TIME seconds **********\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
